lib/srv/db/common/auth.go

```diff
@@ -347,22 +360,58 @@ func (a *dbAuth) GetTLSConfig(ctx context.Context, sessionCtx *Session) (*tls.Co
 	if sessionCtx.Database.IsRDS() || sessionCtx.Database.IsRedshift() || sessionCtx.Database.IsCloudSQL() || sessionCtx.Database.IsAzure() {
 		return tlsConfig, nil
 	}
-	// Otherwise, when connecting to an onprem database, generate a client
-	// certificate. The database instance should be configured with
-	// Teleport's CA obtained with 'tctl auth sign --type=db'.
-	cert, cas, err := a.getClientCert(ctx, sessionCtx)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-	tlsConfig.Certificates = []tls.Certificate{*cert}
-	for _, ca := range cas {
-		if !tlsConfig.RootCAs.AppendCertsFromPEM(ca) {
-			return nil, trace.BadParameter("failed to append CA certificate to the pool")
+
+	// If the root CA has been already added. Skip the client certificate generation.
+	if !rootsAdded {
+		// Otherwise, when connecting to an onprem database, generate a client
+		// certificate. The database instance should be configured with
+		// Teleport's CA obtained with 'tctl auth sign --type=db'.
+		cert, cas, err := a.getClientCert(ctx, sessionCtx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		tlsConfig.Certificates = []tls.Certificate{*cert}
+		for _, ca := range cas {
+			if !tlsConfig.RootCAs.AppendCertsFromPEM(ca) {
+				return nil, trace.BadParameter("failed to append CA certificate to the pool")
+			}
 		}
 	}
+
+	if dbTLSConfig.Mode == types.DatabaseTLSMode_INSECURE {
+		// Accept any certificate provided by database.
+		tlsConfig.InsecureSkipVerify = true
+	} else if dbTLSConfig.Mode == types.DatabaseTLSMode_VERIFY_CA {
+		// Base on https://github.com/golang/go/blob/master/src/crypto/tls/example_test.go#L193-L208
+		// Set InsecureSkipVerify to skip the default validation we are
+		// replacing. This will not disable VerifyConnection.
+		tlsConfig.InsecureSkipVerify = true
+		tlsConfig.VerifyConnection = VerifyConnectionFunc(tlsConfig.RootCAs, dbTLSConfig.ServerName)
```

Comment:
Is there any difference between `verify-ca` and `verify-full` if ServerName is specified? Just wondering if we should be taking ServerName into account at all if the user requested `verify-ca` method.

Reply:
Yeah I understand this is how it's implemented now. My question was rather, _should we_ take server name into account at all if the requested TLS verification method is "verify-ca"? I would think that server name should be irrelevant in that case since we're only going to validate that the certificate is signed by a proper authority.

Reply:
Maybe we can also add a check to return configuration error if `server_name` is not empty when TLS mode is set to `verify-ca`.

Reply:
@r0mant I like the idea of returning the error if `server_name` is not empty in `verify-ca` mode. I'd go even further in this case and also return an error if `server_name` is provided in `insecure` mode (basically only allow `server_name` only in `verify-full`).